File: src/owlet_auth.py
import requests
import json
from typing import Dict, List, Optional, Any
import time

class OwletAuth:
    def __init__(self):
        self.firebase_api_key = 'AIzaSyCx17leGPCKu5tZ1BLPni5LbAAlVvnNxZQ'
        self.firebase_url = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty'
        self.devices_url = 'https://devices-public.owletdata.com/v2'
        self.kms_url = 'https://camera-kms.owletdata.com'
        
        self.jwt_token = None
        self.firebase_refresh_token = None
        self.token_expiry = 0
        self.account_id = None

    async def login(self, email: str, password: str) -> str:
        """Authenticate with Firebase and get JWT token"""
        try:
            print(f"Login attempt - Email: {email}, Password length: {len(str(password))}")
            
            request_data = {
                'email': email,
                'password': password,
                'returnSecureToken': True
            }
            
            print('Firebase request data:', json.dumps({
                'email': request_data['email'],
                'password': f'[{len(request_data["password"])} chars]',
                'returnSecureToken': request_data['returnSecureToken']
            }))
            
            response = requests.post(
                f'{self.firebase_url}/verifyPassword?key={self.firebase_api_key}',
                json=request_data,
                headers={'Content-Type': 'application/json'},
                timeout=10
            )
            
            response.raise_for_status()
            data = response.json()
            
            id_token = data['idToken']
            refresh_token = data['refreshToken']
            expires_in = int(data['expiresIn'])
            local_id = data['localId']
            
            self.jwt_token = id_token
            self.firebase_refresh_token = refresh_token
            self.account_id = local_id
            self.token_expiry = time.time() + expires_in

            print('Successfully authenticated with Firebase')
            return id_token

        except requests.exceptions.HTTPError as e:
            print(f'Firebase authentication failed: {e}')
            
            # Handle specific Firebase authentication errors
            if e.response.status_code == 400:
                try:
                    error_data = e.response.json()
                    firebase_error = error_data.get('error', {})
                    error_code = firebase_error.get('code')
                    
                    if error_code == 'auth/user-not-found':
                        raise Exception('Invalid email address. Please check your Owlet account email.')
                    elif error_code == 'auth/wrong-password':
                        raise Exception('Invalid password. Please check your Owlet account password.')
                    elif error_code == 'auth/invalid-email':
                        raise Exception('Invalid email format. Please enter a valid email address.')
                    elif error_code == 'auth/user-disabled':
                        raise Exception('This Owlet account has been disabled. Please contact Owlet support.')
                    elif error_code == 'auth/too-many-requests':
                        raise Exception('Too many failed login attempts. Please try again later.')
                    else:
                        raise Exception(f'Authentication failed: {firebase_error.get("message", "Unknown error")}')
                except json.JSONDecodeError:
                    raise Exception('Authentication failed: Invalid response from server')
            
            raise Exception(f'Authentication failed: {e}')
            
        except requests.exceptions.ConnectionError:
            raise Exception('Network error: Unable to connect to Owlet servers. Please check your internet connection.')
        except requests.exceptions.Timeout:
            raise Exception('Connection timeout: Owlet servers are not responding. Please try again later.')
        except Exception as e:
            raise Exception(f'Authentication failed: {str(e)}')

    async def refresh_token(self, jwt: str) -> str:
        """Refresh JWT token using Firebase refresh token"""
        if not self.firebase_refresh_token:
            raise Exception('No refresh token available')

        try:
            response = requests.post(
                f'{self.firebase_url}/token?key={self.firebase_api_key}',
                json={
                    'grant_type': 'refresh_token',
                    'refresh_token': self.firebase_refresh_token
                },
                headers={'Content-Type': 'application/json'},
                timeout=10
            )
            
            response.raise_for_status()
            data = response.json()
            
            id_token = data['id_token']
            refresh_token = data['refresh_token']
            expires_in = int(data['expires_in'])
            
            self.jwt_token = id_token
            self.firebase_refresh_token = refresh_token
            self.token_expiry = time.time() + expires_in

            print('Successfully refreshed JWT token')
            return id_token

        except requests.exceptions.HTTPError as e:
            print(f'Token refresh failed: {e}')
            
            # Handle specific Firebase refresh errors
            if e.response.status_code == 400:
                try:
                    error_data = e.response.json()
                    firebase_error = error_data.get('error', {})
                    error_code = firebase_error.get('code')
                    
                    if error_code == 'auth/invalid-grant':
                        raise Exception('Refresh token expired. Please re-authenticate with your Owlet credentials.')
                    elif error_code == 'auth/token-expired':
                        raise Exception('Authentication token expired. Please re-authenticate.')
                    else:
                        raise Exception(f'Token refresh failed: {firebase_error.get("message", "Unknown error")}')
                except json.JSONDecodeError:
                    raise Exception('Token refresh failed: Invalid response from server')
            
            raise Exception(f'Token refresh failed: {e}')
            
        except requests.exceptions.ConnectionError:
            raise Exception('Network error: Unable to refresh authentication token. Please check your internet connection.')
        except Exception as e:
            raise Exception(f'Token refresh failed: {str(e)}')

    async def get_devices(self, jwt: str, account_id: str) -> List[Dict[str, Any]]:
        """Get list of devices from Owlet API"""
        try:
            # Only try the main devices endpoint to avoid rate limiting
            url = f'{self.devices_url}/accounts/{account_id}/devices'
            
            print('Trying endpoint:', url)
            response = requests.get(
                url,
                headers={
                    'Authorization': f'Bearer {jwt}',
                    'Content-Type': 'application/json'
                },
                timeout=10
            )
            
            response.raise_for_status()
            response_data = response.json()
            
            print(f'Response from {url}:', json.dumps(response_data, indent=2))
            
            # Handle different response structures
            devices = response_data.get('devices') or response_data.get('cameras') or response_data or []
            
            if isinstance(devices, list):
                print(f'Found {len(devices)} devices from {url}')
                
                # Log device types
                for index, device in enumerate(devices):
                    print(f'Device {index}: ID={device.get("id")}, Type={device.get("type")}, Name={device.get("name")}')
                
                # Filter for cameras
                cameras = [device for device in devices if 
                          device.get('type') == 'camera' or 
                          device.get('deviceType') == 'CAMERA' or
                          device.get('id', '').startswith('OCA') or
                          'camera' in device.get('id', '').lower()]
                
                print(f'Found {len(cameras)} cameras from {url}')
                print('Device types:', [f"{d.get('id')}:{d.get('type')}" for d in cameras])

                return cameras

            print('No devices found or invalid response format')
            return []

        except requests.exceptions.HTTPError as e:
            print(f'Failed to get devices: {e}')
            
            # Handle specific HTTP status codes
            if e.response.status_code == 401:
                raise Exception('Authentication expired. Please check your Owlet credentials and try again.')
            elif e.response.status_code == 403:
                raise Exception('Access denied. Your Owlet account may not have permission to access cameras.')
            elif e.response.status_code == 404:
                raise Exception('Owlet API endpoint not found. The service may be temporarily unavailable.')
            elif e.response.status_code >= 500:
                raise Exception('Owlet servers are experiencing issues. Please try again later.')
            
            raise Exception(f'Failed to get devices: {e}')
            
        except requests.exceptions.ConnectionError:
            raise Exception('Network error: Unable to connect to Owlet servers. Please check your internet connection.')
        except requests.exceptions.Timeout:
            raise Exception('Connection timeout: Owlet servers are not responding. Please try again later.')
        except Exception as e:
            raise Exception(f'Failed to get devices: {str(e)}')

    async def get_tutk_credentials(self, jwt: str, device_id: str) -> Dict[str, str]:
        """Get TUTK credentials for camera streaming"""
        try:
            response = requests.get(
                f'{self.kms_url}/kms/{device_id}',
                headers={
                    'Authorization': jwt,  # Note: NOT Bearer, just raw token
                    'Content-Type': 'application/json'
                },
                timeout=10
            )
            
            response.raise_for_status()
            return response.json()

        except requests.exceptions.HTTPError as e:
            print(f'Failed to get TUTK credentials for device {device_id}: {e}')
            
            # Handle specific HTTP status codes
            if e.response.status_code == 401:
                raise Exception('Authentication expired. Please check your Owlet credentials and try again.')
            elif e.response.status_code == 403:
                raise Exception('Access denied. Your Owlet account may not have permission to access this camera.')
            elif e.response.status_code == 404:
                raise Exception(f'Camera {device_id} not found. The camera may be offline or removed from your account.')
            elif e.response.status_code >= 500:
                raise Exception('Owlet servers are experiencing issues. Please try again later.')
            
            raise Exception(f'Failed to get TUTK credentials for camera {device_id}: {e}')
            
        except requests.exceptions.ConnectionError:
            raise Exception('Network error: Unable to connect to Owlet servers. Please check your internet connection.')
        except requests.exceptions.Timeout:
            raise Exception('Connection timeout: Owlet servers are not responding. Please try again later.')
        except Exception as e:
            raise Exception(f'Failed to get TUTK credentials for camera {device_id}: {str(e)}')

    def _is_token_valid(self) -> bool:
        """Check if current token is valid"""
        if not self.jwt_token:
            return False
        
        # Check if token expires within the next 5 minutes
        five_minutes_from_now = time.time() + (5 * 60)
        return time.time() < self.token_expiry and self.token_expiry > five_minutes_from_now

    async def ensure_valid_token(self) -> str:
        """Ensure we have a valid token, refresh if needed"""
        if not self._is_token_valid():
            print('JWT token expired or invalid, attempting refresh...')
            try:
                if self.firebase_refresh_token:
                    return await self.refresh_token(self.jwt_token or '')
                else:
                    raise Exception('No refresh token available')
            except Exception as e:
                print(f'Token refresh failed: {e}')
                raise Exception('Authentication expired. Please re-authenticate with your Owlet credentials.')
        
        return self.jwt_token

    def is_authenticated(self) -> bool:
        """Check if currently authenticated"""
        return self._is_token_valid()

    def logout(self):
        """Clear authentication state"""
        self.jwt_token = None
        self.firebase_refresh_token = None
        self.token_expiry = 0
        self.account_id = None

    # Getters for current state
    def get_current_jwt(self) -> Optional[str]:
        """Get current JWT token"""
        return self.jwt_token

    def get_account_id(self) -> Optional[str]:
        """Get current account ID"""
        return self.account_id

    # Legacy methods for backward compatibility
    async def authenticate(self, username: str, password: str):
        """Legacy method for backward compatibility"""
        await self.login(username, password)

    async def get_devices_legacy(self) -> List[Dict[str, Any]]:
        """Legacy method for backward compatibility"""
        if not self.jwt_token or not self.account_id:
            raise Exception('Not authenticated')
        return await self.get_devices(self.jwt_token, self.account_id)

    async def get_camera_connection_info(self, device_id: str) -> Optional[Dict[str, Any]]:
        """Legacy method for backward compatibility"""
        if not self.jwt_token:
            raise Exception('Not authenticated')

        try:
            tutk_credentials = await self.get_tutk_credentials(self.jwt_token, device_id)
            
            # Convert TUTK credentials to legacy format
            return {
                'uid': tutk_credentials['tutkid'],
                'password': tutk_credentials['password'],
                'ipAddress': '',  # Not available from KMS
                'port': 0,  # Not available from KMS
                'deviceId': device_id,
                'status': 'online'
            }
        except Exception as e:
            print(f'Failed to get connection info for device {device_id}: {e}')
            return None

File: src/owlet-auth.ts
import axios from 'axios';

// Firebase Authentication Response
export interface FirebaseAuthResponse {
    kind: string;
    localId: string;
    email: string;
    displayName: string;
    idToken: string;
    registered: boolean;
    refreshToken: string;
    expiresIn: string;
}

// Owlet Device from API
export interface OwletDevice {
    id: string;
    name: string;
    type: string;
    deviceType?: string;
    status: string;
    serialNumber?: string;
    firmwareVersion?: string;
    lastSeen?: string;
    properties?: {
        [key: string]: any;
    };
}

// TUTK Credentials from KMS
export interface TUTKCredentials {
    password: string;
    tutkid: string;
    authKey: string;
}

// Legacy interfaces for backward compatibility
export interface OwletApiResponse<T = any> {
    success: boolean;
    data?: T;
    message?: string;
    error?: string;
}

export interface OwletCameraConnectionInfo {
    uid: string;
    password: string;
    ipAddress: string;
    port: number;
    deviceId: string;
    status: string;
}

export class OwletAuth {
    private firebaseApiKey: string = 'AIzaSyCx17leGPCKu5tZ1BLPni5LbAAlVvnNxZQ';
    private firebaseUrl: string = 'https://www.googleapis.com/identitytoolkit/v3/relyingparty';
    private devicesUrl: string = 'https://devices-public.owletdata.com/v2';
    private kmsUrl: string = 'https://camera-kms.owletdata.com';
    
    private jwtToken: string | null = null;
    private firebaseRefreshToken: string | null = null;
    private tokenExpiry: number = 0;
    private accountId: string | null = null;

    constructor() {
        // No need for axios instance since we're using different endpoints
    }

    /**
     * Authenticate with Firebase and get JWT token
     */
    async login(email: string, password: string): Promise<string> {
        try {
            console.log(`Login attempt - Email: ${email}, Password length: ${password.length}, First char: '${password.charAt(0)}', Last char: '${password.charAt(password.length - 1)}'`);
            
            const requestData = {
                email,
                password,
                returnSecureToken: true
            };
            
            console.log('Firebase request data:', JSON.stringify({
                email: requestData.email,
                password: `[${requestData.password.length} chars]`,
                returnSecureToken: requestData.returnSecureToken
            }));
            
            const response = await axios.post<FirebaseAuthResponse>(
                `${this.firebaseUrl}/verifyPassword?key=${this.firebaseApiKey}`,
                requestData,
                {
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    timeout: 10000
                }
            );

            const { idToken, refreshToken, expiresIn, localId } = response.data;
            
            this.jwtToken = idToken;
            this.firebaseRefreshToken = refreshToken;
            this.accountId = localId;
            this.tokenExpiry = Date.now() + (parseInt(expiresIn) * 1000);

            console.log('Successfully authenticated with Firebase');
            return idToken;

        } catch (error: any) {
            console.error('Firebase authentication failed:', error);
            
            // Handle specific Firebase authentication errors
            if (error.response?.data?.error) {
                const firebaseError = error.response.data.error;
                switch (firebaseError.code) {
                    case 'auth/user-not-found':
                        throw new Error('Invalid email address. Please check your Owlet account email.');
                    case 'auth/wrong-password':
                        throw new Error('Invalid password. Please check your Owlet account password.');
                    case 'auth/invalid-email':
                        throw new Error('Invalid email format. Please enter a valid email address.');
                    case 'auth/user-disabled':
                        throw new Error('This Owlet account has been disabled. Please contact Owlet support.');
                    case 'auth/too-many-requests':
                        throw new Error('Too many failed login attempts. Please try again later.');
                    default:
                        throw new Error(`Authentication failed: ${firebaseError.message}`);
                }
            }
            
            // Handle network errors
            if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
                throw new Error('Network error: Unable to connect to Owlet servers. Please check your internet connection.');
            }
            
            if (error.code === 'ETIMEDOUT') {
                throw new Error('Connection timeout: Owlet servers are not responding. Please try again later.');
            }
            
            throw new Error(`Authentication failed: ${error.message}`);
        }
    }

    /**
     * Refresh JWT token using Firebase refresh token
     */
    async refreshToken(jwt: string): Promise<string> {
        if (!this.firebaseRefreshToken) {
            throw new Error('No refresh token available');
        }

        try {
            const response = await axios.post<{
                id_token: string;
                refresh_token: string;
                expires_in: string;
            }>(
                `${this.firebaseUrl}/token?key=${this.firebaseApiKey}`,
                {
                    grant_type: 'refresh_token',
                    refresh_token: this.firebaseRefreshToken
                },
                {
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    timeout: 10000
                }
            );

            const { id_token, refresh_token, expires_in } = response.data;
            
            this.jwtToken = id_token;
            this.firebaseRefreshToken = refresh_token;
            this.tokenExpiry = Date.now() + (parseInt(expires_in) * 1000);

            console.log('Successfully refreshed JWT token');
            return id_token;

        } catch (error: any) {
            console.error('Token refresh failed:', error);
            
            // Handle specific Firebase refresh errors
            if (error.response?.data?.error) {
                const firebaseError = error.response.data.error;
                switch (firebaseError.code) {
                    case 'auth/invalid-grant':
                        throw new Error('Refresh token expired. Please re-authenticate with your Owlet credentials.');
                    case 'auth/token-expired':
                        throw new Error('Authentication token expired. Please re-authenticate.');
                    default:
                        throw new Error(`Token refresh failed: ${firebaseError.message}`);
                }
            }
            
            // Handle network errors
            if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
                throw new Error('Network error: Unable to refresh authentication token. Please check your internet connection.');
            }
            
            throw new Error(`Token refresh failed: ${error.message}`);
        }
    }

    /**
     * Get list of devices from Owlet API
     */
    async getDevices(jwt: string, accountId: string): Promise<OwletDevice[]> {
        try {
            // Only try the main devices endpoint to avoid rate limiting
            const url = `${this.devicesUrl}/accounts/${accountId}/devices`;
            
            console.log('Trying endpoint:', url);
            const response = await axios.get(
                url,
                {
                    headers: {
                        'Authorization': `Bearer ${jwt}`,
                        'Content-Type': 'application/json'
                    },
                    timeout: 10000
                }
            );

            console.log(`Response from ${url}:`, JSON.stringify(response.data, null, 2));
            
            // Handle different response structures
            const responseData = response.data as any;
            const devices = responseData?.devices || responseData?.cameras || responseData || [];
            
            if (Array.isArray(devices)) {
                console.log(`Found ${devices.length} devices from ${url}`);
                
                // Log device types
                devices.forEach((device: any, index: number) => {
                    console.log(`Device ${index}: ID=${device.id}, Type=${device.type}, Name=${device.name}`);
                });
                
                // Filter for cameras
                const cameras = devices.filter((device: any) => 
                    device.type === 'camera' || 
                    device.deviceType === 'CAMERA' ||
                    device.id?.startsWith('OCA') ||
                    device.id?.includes('camera')
                );
                
                console.log(`Found ${cameras.length} cameras from ${url}`);
                console.log('Device types:', cameras.map(d => `${d.id}:${d.type}`));

                return cameras;
            }

            console.log('No devices found or invalid response format');
            return [];

        } catch (error: any) {
            console.error('Failed to get devices:', error);
            
            // Handle specific HTTP status codes
            if (error.response?.status === 401) {
                throw new Error('Authentication expired. Please check your Owlet credentials and try again.');
            }
            
            if (error.response?.status === 403) {
                throw new Error('Access denied. Your Owlet account may not have permission to access cameras.');
            }
            
            if (error.response?.status === 404) {
                throw new Error('Owlet API endpoint not found. The service may be temporarily unavailable.');
            }
            
            if (error.response?.status >= 500) {
                throw new Error('Owlet servers are experiencing issues. Please try again later.');
            }
            
            // Handle network errors
            if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
                throw new Error('Network error: Unable to connect to Owlet servers. Please check your internet connection.');
            }
            
            if (error.code === 'ETIMEDOUT') {
                throw new Error('Connection timeout: Owlet servers are not responding. Please try again later.');
            }
            
            throw new Error(`Failed to get devices: ${error.message}`);
        }
    }


    /**
     * Get TUTK credentials for camera streaming
     */
    async getTUTKCredentials(jwt: string, deviceId: string): Promise<TUTKCredentials> {
        try {
            const response = await axios.get<TUTKCredentials>(
                `${this.kmsUrl}/kms/${deviceId}`,
                {
                    headers: {
                        'Authorization': jwt, // Note: NOT Bearer, just raw token
                        'Content-Type': 'application/json'
                    },
                    timeout: 10000
                }
            );

            return response.data;

        } catch (error: any) {
            console.error(`Failed to get TUTK credentials for device ${deviceId}:`, error);
            
            // Handle specific HTTP status codes
            if (error.response?.status === 401) {
                throw new Error('Authentication expired. Please check your Owlet credentials and try again.');
            }
            
            if (error.response?.status === 403) {
                throw new Error('Access denied. Your Owlet account may not have permission to access this camera.');
            }
            
            if (error.response?.status === 404) {
                throw new Error(`Camera ${deviceId} not found. The camera may be offline or removed from your account.`);
            }
            
            if (error.response?.status >= 500) {
                throw new Error('Owlet servers are experiencing issues. Please try again later.');
            }
            
            // Handle network errors
            if (error.code === 'ECONNREFUSED' || error.code === 'ENOTFOUND') {
                throw new Error('Network error: Unable to connect to Owlet servers. Please check your internet connection.');
            }
            
            if (error.code === 'ETIMEDOUT') {
                throw new Error('Connection timeout: Owlet servers are not responding. Please try again later.');
            }
            
            throw new Error(`Failed to get TUTK credentials for camera ${deviceId}: ${error.message}`);
        }
    }

    /**
     * Legacy method for backward compatibility
     */
    async authenticate(username: string, password: string): Promise<void> {
        await this.login(username, password);
    }

    /**
     * Legacy method for backward compatibility
     */
    async getDevicesLegacy(): Promise<OwletDevice[]> {
        if (!this.jwtToken || !this.accountId) {
            throw new Error('Not authenticated');
        }
        return this.getDevices(this.jwtToken, this.accountId);
    }

    /**
     * Legacy method for backward compatibility
     */
    async getCameraConnectionInfo(deviceId: string): Promise<OwletCameraConnectionInfo | null> {
        if (!this.jwtToken) {
            throw new Error('Not authenticated');
        }

        try {
            const tutkCredentials = await this.getTUTKCredentials(this.jwtToken, deviceId);
            
            // Convert TUTK credentials to legacy format
            return {
                uid: tutkCredentials.tutkid,
                password: tutkCredentials.password,
                ipAddress: '', // Not available from KMS
                port: 0, // Not available from KMS
                deviceId,
                status: 'online'
            };
        } catch (error: any) {
            console.error(`Failed to get connection info for device ${deviceId}:`, error);
            return null;
        }
    }

    private isTokenValid(): boolean {
        if (!this.jwtToken) {
            return false;
        }
        
        // Check if token expires within the next 5 minutes
        const fiveMinutesFromNow = Date.now() + (5 * 60 * 1000);
        return Date.now() < this.tokenExpiry && this.tokenExpiry > fiveMinutesFromNow;
    }

    async ensureValidToken(): Promise<string> {
        if (!this.isTokenValid()) {
            console.log('JWT token expired or invalid, attempting refresh...');
            try {
                if (this.firebaseRefreshToken) {
                    return await this.refreshToken(this.jwtToken || '');
                } else {
                    throw new Error('No refresh token available');
                }
            } catch (error: any) {
                console.error('Token refresh failed:', error);
                throw new Error('Authentication expired. Please re-authenticate with your Owlet credentials.');
            }
        }
        
        return this.jwtToken!;
    }

    isAuthenticated(): boolean {
        return this.isTokenValid();
    }

    logout(): void {
        this.jwtToken = null;
        this.firebaseRefreshToken = null;
        this.tokenExpiry = 0;
        this.accountId = null;
    }

    // Getters for current state
    getCurrentJWT(): string | null {
        return this.jwtToken;
    }

    getAccountId(): string | null {
        return this.accountId;
    }
}
